bemani/backend/iidx/rootage.py

```diff
@@ -0,0 +1,1949 @@
+            13: 70,
+            14: 75,
+            15: 75,
+            16: 80,
+            17: 80,
+            18: 80,
+        }
+        for item in request.children:
+            if item.name == 'music_list':
+                music_list[int(item.child_value('index'))] = int(item.child_value('total_notes'))
+            if item.name == 'cpu_list':
+                cpu_list[int(item.child_value('index'))] = int(item.child_value('grade_id'))
+        for index in range(4):
+            cpu_score_list = Node.void('cpu_score_list')
+            cpu_score_list.add_child(Node.s32('index', index))
+            root.add_child(cpu_score_list)
+            for i in range(4):
+                score_list = Node.void('score_list')
+                cpu_score_list.add_child(score_list)
+                percentage = grade_to_percentage_map.get(cpu_list[index], 0)
+                notes = music_list[i]
+                score = (notes * 2 * percentage) // 100
+                ghost_section = score // 64
+                ghost = [ghost_section] * 64
+                ghost_total = ghost_section * 64
+                ghost_lost = score - ghost_total
+                ghost[0] = ghost[0] + ghost_lost
+                score_list.add_child(Node.s32('score', score))
+                score_list.add_child(Node.bool('enable_score', True))
+                score_list.add_child(Node.u8_array('ghost', ghost))
+                score_list.add_child(Node.bool('enable_ghost', True))
+        return root
+
+    def handle_IIDX26grade_raised_request(self, request: Node) -> Node:
+        extid = int(request.attribute('iidxid'))
+        cltype = int(request.attribute('gtype'))
+        rank = self.game_to_db_rank(int(request.attribute('gid')), cltype)
+
+        userid = self.data.remote.user.from_extid(self.game, self.version, extid)
+        if userid is not None:
+            percent = int(request.attribute('achi'))
+            stages_cleared = int(request.attribute('cstage'))
+            cleared = stages_cleared == self.DAN_STAGES
+
+            if cltype == self.GAME_CLTYPE_SINGLE:
+                index = self.DAN_RANKING_SINGLE
+            else:
+                index = self.DAN_RANKING_DOUBLE
+
+            self.update_rank(
+                userid,
+                index,
+                rank,
+                percent,
+                cleared,
+                stages_cleared,
+            )
+
+        # Figure out number of players that played this ranking
+        all_achievements = self.data.local.user.get_all_achievements(self.game, self.version)
+        num_players = 0
+        for [_, ach] in all_achievements:
+            if ach.type != index:
+                continue
+            if ach.id != rank:
+                continue
+            num_players = num_players + 1
+
+        root = Node.void('IIDX26grade')
+        root.set_attribute('pnum', str(num_players))
+        return root
+
+    def handle_IIDX26pc_common_request(self, request: Node) -> Node:
```

Comment:
I think you mentioned Rootage and above don't have courses anymore?

Reply:
Yes, interestingly ranking.getranker is still called but it's a no op pretty much now